# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/base/Pivot.cpp
# Commit: 892b785f0fc5
# Full Hash: 892b785f0fc5126b67898f2534789160b89e8744
# Author: Morgan Rae Reschenberg <mreschenberg@mozilla.com>
# Date: 2022-10-28 04:43:55
# Regressor Bug: 1787284
# File Overlap Count: 1
# Description:
#   Bug 1787284: [Part 2] Cache name and create radio name pivot rule r=eeejay
#   
#   Depends on D158892
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D159118
# ==============================================================================

diff -r d3d8e4d17161 -r 892b785f0fc5 accessible/base/Pivot.cpp
--- a/accessible/base/Pivot.cpp	Thu Oct 27 20:32:17 2022 +0000
+++ b/accessible/base/Pivot.cpp	Thu Oct 27 20:32:18 2022 +0000
@@ -634,3 +634,29 @@
   }
   return nsIAccessibleTraversalRule::FILTER_MATCH;
 }
+
+// Radio Button Name Rule
+
+PivotRadioNameRule::PivotRadioNameRule(const nsString& aName) : mName(aName) {}
+
+uint16_t PivotRadioNameRule::Match(Accessible* aAcc) {
+  uint16_t result = nsIAccessibleTraversalRule::FILTER_IGNORE;
+  RemoteAccessible* remote = aAcc->AsRemote();
+  if (!remote || !StaticPrefs::accessibility_cache_enabled_AtStartup()) {
+    // We need the cache to be able to fetch the name attribute below.
+    return result;
+  }
+
+  if (nsAccUtils::MustPrune(aAcc) || aAcc->IsOuterDoc()) {
+    result |= nsIAccessibleTraversalRule::FILTER_IGNORE_SUBTREE;
+  }
+
+  if (remote->IsHTMLRadioButton()) {
+    nsString currName = remote->GetCachedHTMLRadioNameAttribute();
+    if (!currName.IsEmpty() && mName.Equals(currName)) {
+      result |= nsIAccessibleTraversalRule::FILTER_MATCH;
+    }
+  }
+
+  return result;
+}